File: Power Query/300-399/379/PQ_379.py
import pandas as pd
from collections import deque

path = "300-399/379/PQ_Challenge_379.xlsx"
input = pd.read_excel(path, usecols="A:D", nrows=21, skiprows=0)
test = pd.read_excel(path, usecols="F:G", nrows=10, skiprows=0).rename(columns=lambda c: c.replace(".1", ""))

q = deque()
out = []

for _, r in input.iterrows():
    if r["Type"] == "BUY":
        q.append([r["Coins"], r["Price"]])
    else:
        need, cost = r["Coins"], 0
        while need:
            qty, price = q[0]
            take = min(qty, need)
            cost += take * price
            q[0][0] -= take
            if q[0][0] == 0:
                q.popleft()
            need -= take
        
        profit = r["Coins"] * r["Price"] - cost
        out.append((r["TransactionID"], profit))

result = pd.DataFrame(out, columns=["TransactionID", "Profit"])
print(result.equals(test))
# True

File: Power Query/300-399/379/PQ_379.R
library(tidyverse)
library(readxl)
library(R6)

path <- "300-399/379/PQ_Challenge_379.xlsx"
input <- read_excel(path, range = "A1:D21")
test <- read_excel(path, range = "F1:G11")

Deque <- R6Class(
  "Deque",
  public = list(
    q = list(),
    push_back = function(x) self$q <- c(self$q, list(x)),
    push_front = function(x) self$q <- c(list(x), self$q),
    pop_front = function() {
      x <- self$q[[1]]
      self$q <- self$q[-1]
      x
    },
    peek = function() self$q[[1]],
    empty = function() length(self$q) == 0
  )
)
fifo <- Deque$new()
res <- input %>%
  mutate(
    Profit = pmap_dbl(list(Type, Coins, Price), \(type, qty, price) {
      if (type == "BUY") {
        fifo$push_back(c(qty, price))
        return(NA_real_)
      }
      need <- qty
      cost <- 0
      while (need > 0) {
        lot <- fifo$pop_front()
        take <- min(lot[1], need)
        cost <- cost + take * lot[2]
        remaining <- lot[1] - take
        if (remaining > 0) {
          fifo$push_front(c(remaining, lot[2]))
        }
        need <- need - take
      }
      qty * price - cost
    })
  ) %>%
  filter(Type == "SELL") %>%
  select(TransactionID, Profit)

all.equal(res, test)
## [1] TRUE
